In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data=pd.read_csv('/content/drive/MyDrive/Datasets MLAI/mobile_price_classification.csv')

In [3]:
data

battery_power  bluetooth  clock_speed  dual_sim  front_cam  4G  \
0               842          0          2.2         0          1   0   
1              1021          1          0.5         1          0   1   
2               563          1          0.5         1          2   1   
3               615          1          2.5         0          0   0   
4              1821          1          1.2         0         13   1   
...             ...        ...          ...       ...        ...  ..   
1995            794          1          0.5         1          0   1   
1996           1965          1          2.6         1          0   0   
1997           1911          0          0.9         1          1   1   
1998           1512          0          0.9         0          4   1   
1999            510          1          2.0         1          5   1   

      int_memory  m_dep  mobile_wt  n_cores  ...  px_height  px_width   ram  \
0              7    0.6        188        2  ...         20       756  2549   
1             53    0.7        136        3  ...        905      1988  2631   
2             41    0.9        145        5  ...       1263      1716  2603   
3             10    0.8        131        6  ...       1216      1786  2769   
4             44    0.6        141        2  ...       1208      1212  1411   
...          ...    ...        ...      ...  ...        ...       ...   ...   
1995           2    0.8        106        6  ...       1222      1890   668   
1996          39    0.2        187        4  ...        915      1965  2032   
1997          36    0.7        108        8  ...        868      1632  3057   
1998          46    0.1        145        5  ...        336       670   869   
1999          45    0.9        168        6  ...        483       754  3919   

      sc_h  sc_w  talk_time  three_g  touch_screen  wifi  price_range  
0        9     7         19        0             0     1            1  
1       17     3          7        1             1     0            2  
2       11     2          9        1             1     0            2  
3       16     8         11        1             0     0            2  
4        8     2         15        1             1     0            1  
...    ...   ...        ...      ...           ...   ...          ...  
1995    13     4         19        1             1     0            0  
1996    11    10         16        1             1     1            2  
1997     9     1          5        1             1     0            3  
1998    18    10         19        1             1     1            0  
1999    19     4          2        1             1     1            3  

[2000 rows x 21 columns]

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   battery_power   2000 non-null   int64  
 1   bluetooth       2000 non-null   int64  
 2   clock_speed     2000 non-null   float64
 3   dual_sim        2000 non-null   int64  
 4   front_cam       2000 non-null   int64  
 5   4G              2000 non-null   int64  
 6   int_memory      2000 non-null   int64  
 7   m_dep           2000 non-null   float64
 8   mobile_wt       2000 non-null   int64  
 9   n_cores         2000 non-null   int64  
 10  primary_camera  2000 non-null   int64  
 11  px_height       2000 non-null   int64  
 12  px_width        2000 non-null   int64  
 13  ram             2000 non-null   int64  
 14  sc_h            2000 non-null   int64  
 15  sc_w            2000 non-null   int64  
 16  talk_time       2000 non-null   int64  
 17  three_g         2000 non-null   i

In [5]:
data.describe()

battery_power  bluetooth  clock_speed     dual_sim    front_cam  \
count    2000.000000  2000.0000  2000.000000  2000.000000  2000.000000   
mean     1238.518500     0.4950     1.522250     0.509500     4.309500   
std       439.418206     0.5001     0.816004     0.500035     4.341444   
min       501.000000     0.0000     0.500000     0.000000     0.000000   
25%       851.750000     0.0000     0.700000     0.000000     1.000000   
50%      1226.000000     0.0000     1.500000     1.000000     3.000000   
75%      1615.250000     1.0000     2.200000     1.000000     7.000000   
max      1998.000000     1.0000     3.000000     1.000000    19.000000   

                4G   int_memory        m_dep    mobile_wt      n_cores  ...  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000  ...   
mean      0.521500    32.046500     0.501750   140.249000     4.520500  ...   
std       0.499662    18.145715     0.288416    35.399655     2.287837  ...   
min       0.000000     2.000000     0.100000    80.000000     1.000000  ...   
25%       0.000000    16.000000     0.200000   109.000000     3.000000  ...   
50%       1.000000    32.000000     0.500000   141.000000     4.000000  ...   
75%       1.000000    48.000000     0.800000   170.000000     7.000000  ...   
max       1.000000    64.000000     1.000000   200.000000     8.000000  ...   

         px_height     px_width          ram         sc_h         sc_w  \
count  2000.000000  2000.000000  2000.000000  2000.000000  2000.000000   
mean    645.108000  1251.515500  2124.213000    12.306500     5.767000   
std     443.780811   432.199447  1084.732044     4.213245     4.356398   
min       0.000000   500.000000   256.000000     5.000000     0.000000   
25%     282.750000   874.750000  1207.500000     9.000000     2.000000   
50%     564.000000  1247.000000  2146.500000    12.000000     5.000000   
75%     947.250000  1633.000000  3064.500000    16.000000     9.000000   
max    1960.000000  1998.000000  3998.000000    19.000000    18.000000   

         talk_time      three_g  touch_screen         wifi  price_range  
count  2000.000000  2000.000000   2000.000000  2000.000000  2000.000000  
mean     11.011000     0.761500      0.503000     0.507000     1.500000  
std       5.463955     0.426273      0.500116     0.500076     1.118314  
min       2.000000     0.000000      0.000000     0.000000     0.000000  
25%       6.000000     1.000000      0.000000     0.000000     0.750000  
50%      11.000000     1.000000      1.000000     1.000000     1.500000  
75%      16.000000     1.000000      1.000000     1.000000     2.250000  
max      20.000000     1.000000      1.000000     1.000000     3.000000  

[8 rows x 21 columns]

In [6]:
data.isna().sum()

battery_power     0
bluetooth         0
clock_speed       0
dual_sim          0
front_cam         0
4G                0
int_memory        0
m_dep             0
mobile_wt         0
n_cores           0
primary_camera    0
px_height         0
px_width          0
ram               0
sc_h              0
sc_w              0
talk_time         0
three_g           0
touch_screen      0
wifi              0
price_range       0
dtype: int64

In [7]:
data['price_range'].value_counts()

1    500
2    500
3    500
0    500
Name: price_range, dtype: int64

**Separating dependent variables and independent variables**

In [8]:
x=data.drop('price_range',axis=1)
y=data['price_range']

In [9]:
x.shape

(2000, 20)

**Standard scaling**

In [10]:
from sklearn.preprocessing import StandardScaler
std=StandardScaler()
x=std.fit_transform(x)
x=pd.DataFrame(x)

**One-Hot Encoding**

In [11]:
y_encoded=pd.get_dummies(y)

**Splitting data into train and test data**

In [12]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y_encoded,test_size=0.2,random_state=42)

# **Artificial Neural Network**

In [13]:
from keras.models import Sequential
from keras.layers import Dense

In [14]:
model=Sequential()

model.add(Dense(units=10,activation='relu'))

model.add(Dense(units=4,activation='softmax'))

In [15]:
model.compile(loss = 'categorical_crossentropy', optimizer = 'adam',metrics = ['accuracy'])

In [16]:
model.fit(x_train,y_train,epochs=100)

Epoch 1/100
50/50 [==============================] - 3s 9ms/step - loss: 1.6708 - accuracy: 0.2706
Epoch 2/100
50/50 [==============================] - 1s 10ms/step - loss: 1.5403 - accuracy: 0.2831
Epoch 3/100
50/50 [==============================] - 0s 4ms/step - loss: 1.4548 - accuracy: 0.3013
Epoch 4/100
50/50 [==============================] - 0s 6ms/step - loss: 1.3924 - accuracy: 0.3350
Epoch 5/100
50/50 [==============================] - 0s 5ms/step - loss: 1.3402 - accuracy: 0.3631
Epoch 6/100
50/50 [==============================] - 0s 6ms/step - loss: 1.2934 - accuracy: 0.3894
Epoch 7/100
50/50 [==============================] - 0s 7ms/step - loss: 1.2467 - accuracy: 0.4281
Epoch 8/100
50/50 [==============================] - 0s 6ms/step - loss: 1.1995 - accuracy: 0.4594
Epoch 9/100
50/50 [==============================] - 0s 8ms/step - loss: 1.1501 - accuracy: 0.4913
Epoch 10/100
50/50 [==============================] - 0s 9ms/step - loss: 1.0993 - accuracy: 0.5331
Epoch 11

In [17]:
y_predictions=model.predict(x_test)

13/13 [==============================] - 0s 2ms/step


In [18]:
y_predictions

array([[9.83092725e-01, 1.68902613e-02, 1.70827407e-05, 6.50389696e-16],
       [4.07274285e-07, 6.19484708e-02, 9.37979460e-01, 7.16720533e-05],
       [1.90591272e-02, 9.79247332e-01, 1.69350579e-03, 1.67984170e-12],
       ...,
       [2.44216353e-04, 9.65630591e-01, 3.41250822e-02, 3.88566832e-08],
       [6.70504814e-04, 9.85385180e-01, 1.39444275e-02, 5.39460254e-10],
       [8.83019838e-11, 1.06896344e-03, 9.91906226e-01, 7.02474546e-03]],
      dtype=float32)

In [19]:
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Accuracy: {accuracy * 100:.2f}%')

13/13 [==============================] - 0s 2ms/step - loss: 0.1304 - accuracy: 0.9650
Accuracy: 96.50%


# **Hyperparameter tuning using Grid Search CV**

In [20]:
!pip install scikeras

In [21]:
def create_model(optimizer='adam'):
  model=Sequential()
  model.add(Dense(units=10,input_shape=(20,),activation='relu'))
  model.add(Dense(units=4,activation='softmax'))
  model.compile(loss = 'categorical_crossentropy', optimizer = 'adam',metrics = ['accuracy'])
  return model

In [22]:
from scikeras.wrappers import KerasClassifier

In [23]:
model=KerasClassifier(model=create_model,batch_size=10)

In [24]:
# define the grid search parameters
optimizer = ['SGD',  'Adam', 'Adamax', 'Nadam']
epochs = [60,80,100]
learn_rate = [0.001, 0.01, 0.1, 0.2, 0.3]
param_grid = dict(model__optimizer=optimizer,optimizer__learning_rate=learn_rate,epochs=epochs)

In [25]:
from sklearn.model_selection import GridSearchCV
grid=GridSearchCV(estimator=model,param_grid=param_grid,n_jobs=-1,cv=3)
grid_result=grid.fit(x_train,y_train)

/usr/local/lib/python3.10/dist-packages/joblib/externals/loky/process_executor.py:752: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
  warnings.warn(


Epoch 1/80
160/160 [==============================] - 1s 2ms/step - loss: 1.5958 - accuracy: 0.2244
Epoch 2/80
160/160 [==============================] - 0s 2ms/step - loss: 1.3599 - accuracy: 0.3094
Epoch 3/80
160/160 [==============================] - 0s 2ms/step - loss: 1.1991 - accuracy: 0.4400
Epoch 4/80
160/160 [==============================] - 0s 2ms/step - loss: 1.0522 - accuracy: 0.5362
Epoch 5/80
160/160 [==============================] - 0s 2ms/step - loss: 0.9177 - accuracy: 0.6231
Epoch 6/80
160/160 [==============================] - 0s 3ms/step - loss: 0.8076 - accuracy: 0.6750
Epoch 7/80
160/160 [==============================] - 0s 3ms/step - loss: 0.7235 - accuracy: 0.7206
Epoch 8/80
160/160 [==============================] - 0s 3ms/step - loss: 0.6586 - accuracy: 0.7544
Epoch 9/80
160/160 [==============================] - 0s 3ms/step - loss: 0.6057 - accuracy: 0.7837
Epoch 10/80
160/160 [==============================] - 0s 3ms/step - loss: 0.5606 - accuracy: 0.8050

In [28]:
grid_result.best_score_

0.9462538618471751

0.9462538618471751

In [29]:
grid_result.best_params_

{'epochs': 80, 'model__optimizer': 'Adamax', 'optimizer__learning_rate': 0.001}

In [31]:
model=Sequential()

model.add(Dense(units=10,activation='relu'))

model.add(Dense(units=4,activation='softmax'))

In [40]:
model.compile(loss = 'categorical_crossentropy', optimizer = 'Adamax',metrics = ['accuracy'])

In [41]:
model.fit(x_train,y_train,epochs=100)

Epoch 1/100
50/50 [==============================] - 1s 3ms/step - loss: 0.3025 - accuracy: 0.9319
Epoch 2/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2973 - accuracy: 0.9325
Epoch 3/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2937 - accuracy: 0.9350
Epoch 4/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2898 - accuracy: 0.9369
Epoch 5/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2863 - accuracy: 0.9369
Epoch 6/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2829 - accuracy: 0.9369
Epoch 7/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2797 - accuracy: 0.9388
Epoch 8/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2767 - accuracy: 0.9400
Epoch 9/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2737 - accuracy: 0.9406
Epoch 10/100
50/50 [==============================] - 0s 3ms/step - loss: 0.2705 - accuracy: 0.9413
Epoch 11/

In [42]:
y_predictions=model.predict(x_test)

13/13 [==============================] - 0s 2ms/step


In [43]:
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Accuracy: {accuracy * 100:.2f}%')

13/13 [==============================] - 0s 3ms/step - loss: 0.1574 - accuracy: 0.9550
Accuracy: 95.50%


**After hypertuning using GridSearchCV, accuracy=95.5%**

# **Hypertuning using Optuna**

In [36]:
pip install optuna tensorflow

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 409.6/409.6 kB 5.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 226.8/226.8 kB 10.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 7.9 MB/s eta 0:00:00


In [43]:
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

In [77]:
def create_model(trial):
    # Define the hyperparameters to be tuned
    units = trial.suggest_int('units', 32, 512, log=True)
    dropout_rate = trial.suggest_uniform('dropout_rate', 0.2, 0.5)
    learning_rate = trial.suggest_loguniform('learning_rate', 0.001, 0.1)
    epochs = trial.suggest_int('epochs', 5, 100)

    # Build the neural network model
    model = Sequential()
    model.add(layers.Dense(units, activation='relu', input_shape=(20,)))
    model.add(layers.Dropout(dropout_rate))
    model.add(layers.Dense(4, activation='softmax'))

    # Compile the model
    optimizer = keras.optimizers.Adam(learning_rate=learning_rate)
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    return model, epochs

In [78]:
def objective(trial):
    # Create the model
    model, epochs= create_model(trial)

    # Train the model
    history = model.fit(x_train, y_train, epochs=epochs, validation_data=(x_test, y_test), verbose=0)

    # Evaluate the model on the validation set
    score = model.evaluate(x_test, y_test, verbose=0)

    # Return the validation accuracy as the objective value to be minimized
    return 1.0 - score[1]

In [81]:
import optuna
# Create an Optuna study object and specify the optimization direction
study = optuna.create_study(direction='minimize')

# Optimize the objective function
study.optimize(objective, n_trials=50)


[I 2023-11-21 15:38:55,071] A new study created in memory with name: no-name-67a3d789-b54f-407b-a61a-94aa1695adfe
<ipython-input-77-abb92030d5ea>:4: FutureWarning: suggest_uniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float instead.
  dropout_rate = trial.suggest_uniform('dropout_rate', 0.2, 0.5)
<ipython-input-77-abb92030d5ea>:5: FutureWarning: suggest_loguniform has been deprecated in v3.0.0. This feature will be removed in v6.0.0. See https://github.com/optuna/optuna/releases/tag/v3.0.0. Use suggest_float(..., log=True) instead.
  learning_rate = trial.suggest_loguniform('learning_rate', 0.001, 0.1)
[I 2023-11-21 15:38:58,388] Trial 0 finished with value: 0.05500000715255737 and parameters: {'units': 79, 'dropout_rate': 0.218852444659196, 'learning_rate': 0.0077910056490298595, 'epochs': 13}. Best is trial 0 with value: 0.05500000715255737.
<ipython-input-77-abb92030d5ea>:4: Future

In [82]:
trial = study.best_trial
print('Loss: {}'.format(trial.value))
print("Best hyperparameters: {}".format(trial.params))

Loss: 0.019999980926513672
Best hyperparameters: {'units': 33, 'dropout_rate': 0.41134606896670484, 'learning_rate': 0.0016544867630980393, 'epochs': 64}


In [ ]:
#After hypertuning

In [83]:
model = Sequential()
model.add(layers.Dense(33, activation='relu', input_shape=(20,)))
model.add(layers.Dropout(0.41134606896670484))
model.add(layers.Dense(4, activation='softmax'))

    # Compile the model
optimizer = keras.optimizers.Adam(learning_rate=0.0016544867630980393)
model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

In [84]:
model.fit(x_train,y_train,epochs=64)

Epoch 1/64
50/50 [==============================] - 1s 2ms/step - loss: 1.6408 - accuracy: 0.2463
Epoch 2/64
50/50 [==============================] - 0s 2ms/step - loss: 1.4041 - accuracy: 0.3475
Epoch 3/64
50/50 [==============================] - 0s 2ms/step - loss: 1.2369 - accuracy: 0.4219
Epoch 4/64
50/50 [==============================] - 0s 2ms/step - loss: 1.1146 - accuracy: 0.4812
Epoch 5/64
50/50 [==============================] - 0s 2ms/step - loss: 0.9937 - accuracy: 0.5606
Epoch 6/64
50/50 [==============================] - 0s 2ms/step - loss: 0.8857 - accuracy: 0.6131
Epoch 7/64
50/50 [==============================] - 0s 2ms/step - loss: 0.8225 - accuracy: 0.6206
Epoch 8/64
50/50 [==============================] - 0s 2ms/step - loss: 0.7534 - accuracy: 0.6725
Epoch 9/64
50/50 [==============================] - 0s 2ms/step - loss: 0.6982 - accuracy: 0.7088
Epoch 10/64
50/50 [==============================] - 0s 2ms/step - loss: 0.6511 - accuracy: 0.7500
Epoch 11/64
50/50 [

In [85]:
loss, accuracy = model.evaluate(x_test, y_test)
print(f'Accuracy: {accuracy * 100:.2f}%')

13/13 [==============================] - 0s 4ms/step - loss: 0.0938 - accuracy: 0.9700
Accuracy: 97.00%


**After hyperparameter tuning using optuna,accuracy=97%**